### PR TITLE
new scheme according to the final SSB proposal from 3.4.2016

### PR DESCRIPTION
**Pull request by ahashibon**
@khiltunen @kemattil @pooyan-dadvand @simphony/cimne @simphony/fraunhofer-iwm @simphony/enthought @SGGgarcia and Guillermo

Please review the new schema according to the SSB proposal. 
It is an initial version, more relations, elements, etc are needed.

### CODE REVIEW

**Comment by tuopuu** [comment]
@ahashibon, there seems to be some overlap between cuba.yml and material_relation_definitions.yml with simphony_metadata.yml. Some definitions are currently present in two of these files. Are we supposed to add to simphony_metadata.yml the contents of material_relation_definitions.yml, so that we can remove that specific file completely. Is this what we're aiming to do?  

Why should we have a separate file cuds_item.yml for low level Cuds data objects? Shouldn't that be also merged with simphony_metadata.yml? Also, what are the final YAML-files we end up using? Just simphony-metadata.yml and cuba.yml?

Because there seems to be so many definitions that need to be added to simphony_metadata.yml, should we (the developers) work in parallel updating this branch before finally merging it to simphony-metadata master? Another option would be to merge this quickly, and then start working individually with separate branches that add new definitions.

**Comment by mehdisadeghi** [comment]
IMO one file for CUDS and one for CUBA are enough. We might end up splitting specification into multiple files in order to maintain them effectively however.

**Comment by tuopuu** [comment]
@mehdisadeghi, please take a look at mine and Adham's comment at #1. I think we can start with just two files for now, and if necessary, later split them into smaller files.

**Comment by mehdisadeghi** [comment]
@tuopuu I agree.

**Comment by ahashibon** [comment]
The goal would be to have one source for all metadata (simphony_metadata.yml), however right now the cuba.yml has very elemental metadata. It will take time to merge them, so my suggestion is also to leave it as is, and comment-out or better ignore those in the simphony_metadata.yml file that are also in cuba.yml for the mean time. The same for CUDS_item items. 

The materials relations should be however moved to the new metadata file, this is the decision of the SSB already at the M24 meeting. We already moved most of them.

**Comment by ahashibon** [comment]
@tuopuu what final YAML-files we end up using? Just simphony-metadata.yml and cuba.yml?

This is another plausible solution, if we define the elementary cuba (in cuba.yml) using a differnet schema (the cuba are the elements that are simple, and are used across the board) while the simphony_metadata are the more complex data objects that are domain specific. This is a matter of ongoing research also in the wider level of the cluster.

**Comment by tuopuu** [comment]
Thanks @ahashibon for clarification. How do we continue to work on this, as it's related to #1? Do you mind if I take over this branch?

**Comment by ahashibon** [comment]
@kemattil @khiltunen @gromansg @pooyan-dadvand @tuopuu @mehdisadeghi can this be merged ?

File: simphony_metadata.yml [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    physics_variables: [CUBA.POSITION, CUBA.POTENTIAL_ENERGY]
+    CUBA.CUTOFF_DISTANCE:
+      default: 1.0
+    CUBA.ENERGY_WELL_DEPTH:
+      default: 1.0
+    CUBA.VAN_DER_WAALS_RADIUS:
+      default: 1.0
+
+  COULOMB:
+  parent: CUBA.PAIR_POTENTIAL
```
Comment (tuopuu) [inline simphony_metadata.yml]: This should be indented.

**Comment by mehdisadeghi** [comment]
> my suggestion is also to leave it as is, and comment-out or better ignore those in the simphony_metadata.yml

I would suggest to merge them into the new files (CUDS & CUBA) and delete the old ones afterwards, otherwise it will be confusing.

**Comment by tuopuu** [comment]
@ahashibon, just one minor comment on the file. I also think that we should pick a convention for the number of empty lines between definitions of the objects: currently there seems to be a variation between 1-3. I propose using always one empty line.

**Comment by tuopuu** [comment]
Some of the definitions seem to have a period at the end. I think we should also agree on a convention that period is only used for definitions that are clearly sentences, and leave the other "header-like" definitions without it.

File: simphony_metadata.yml [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+      default: 1e-3
+    CUBA.RELAXATION_TIME: 
+      default:1
+    CUBA.LINEAR_CONSTANT:
+      default: 1e-5
+    CUBA.POWER_LAW_INDEX:
+      default: 1
+
+  CROSS_POWER_LAW_MODEL:
+    definitoin: Viscosity Cross power law model
```
Comment (tuopuu) [inline simphony_metadata.yml]: this should be definition.

**Comment by kemattil** [comment]
Please do not merge yet. I have some comments, put not yet time to list them.
Will do it tomorrow or Friday at the latest ...

**Comment by ahashibon** [comment]
@tuopuu Thanks, good comments, please check now if the dots and spaces are ok.

**Comment by ahashibon** [comment]
few more minor formatting.

**Comment by ahashibon** [comment]
closed by mistake! sorry about that. Working through a very slow network.

File: simphony_metadata.yml [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+  GRANULAR_DYNAMICS:
+    definition: Granular dynamics of spherical particles using DEM
+    parent: CUBA.PHYSICS_EQUATION    
+    models: [CUBA.MESCOCOPIC]
+    physics_variables: [CUBA.POSITION, CUBA.VELOCITY, CUBA.MOMENTUM, CUBA.ACCELERATION, CUBA.MOMENT_INERTIA, CUBA.TORQUE, CUBA.ANGULAR_VELOCITY]  
+    
+  CFD:
+    definition: Computational fluid dynamics general (set of ) equations for momentum, mass and energy
+    parent": CUBA.PHYSICS_EQUATION
```
Comment (tuopuu) [inline simphony_metadata.yml]: Remove "

File: simphony_metadata.yml [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+      scope: CUBA.USER
+    CUBA.TURBULENCE_MODEL:
+      default: CUBA.LAMINAR_FLOW_MODEL
+    CUBA.MULTIPHASE_MODEL: 
+      default: CUBA.SINGLE_PHASE_MODEL
+    CUBA.RHEOLOGY_MODEL:
+      default: CUBA.NEWTONIAN_MODEL
+    CUBA.GRAVITY_MODEL:
+    CUBA.ELECTROSTATIC_MODEL:
+      defualt: CUBA.CONSTANT_ELECTROSTATIC_FIELD_MODEL
```
Comment (tuopuu) [inline simphony_metadata.yml]: Change to default

File: simphony_metadata.yml [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+    CUBA.ACCELERATION:
+      default: [0, 0, 0]
+
+  CONSTANT_ELECTROSTATIC_FIELD_MODEL:
+    definition: A constant elecrostatic field model 
+    parent: CUBA.PHYSICS_EQUATION
+    models: [CUBA.MESOSCOPIC, CUBA.CONTINUUM]
+    physics_variables: [CUBA.ELECTRIC_FIELD, CUBA.CHARGE]
+    CUBA.ELECTROSTATIC_FIELD:
+      default: [0, 0, 0.0]
```
Comment (tuopuu) [inline simphony_metadata.yml]: Could use just [0, 0, 0] here.

File: simphony_metadata.yml [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+    definition: Non Isothermal heat transport model
+    parent: CUBA.PHYSICS_EQUATION
+    models: [CUBA.CONTINUUM]
+    physics_quantities: [CUBA.TEMPERATURE, CUBA.HEAT_CONDUCTIVITY]
+      
+  INCOMPRESSIBLE_FLUID_MODEL:
+    parent: CUBA.PHYSICS_EQUATION
+    models: [CUBA.CONTINUUM]
+    definition: Incompressible fluid model
+    physics_quantities: [CUBA.VELOCITY, CUBA.POSITION, CUBA.DENSITY, CUBA.VICOSITY]
```
Comment (tuopuu) [inline simphony_metadata.yml]: Change the last key to CUBA.VISCOSITY

File: simphony_metadata.yml [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+    
+  LAMINAR_FLOW_MODEL:
+    parent: CUBA.PHYSICS_EQUATION
+    definition: Laminar model.
```
Comment (tuopuu) [inline simphony_metadata.yml]: Remove period

File: simphony_metadata.yml [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+    models: [CUBA.CONTINUUM]
+  
+  POWER_LAW_VISCOSITY_MODEL:
+    definition: Power Law model for a variable viscosity function that is limited by minimum and maximum values 
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.MAXIMUM_VISCOSITY: 
+      default: 1e-3
+    CUBA.MINIMUM_VISCOSITY:
+      DEFAULT:1E-5
```
Comment (tuopuu) [inline simphony_metadata.yml]: change to lower case: default: 1e-5

File: simphony_metadata.yml [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+      default: 1
+      
+  HERSCHEL_BULKLEY_MODEL:
+    definition: Herschel-Bulkley model combines the effects of Bingham plastic and power-law behavior in a fluid
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.INITIAL_VISCOSITY: 
+      default: 1e-3
+    CUBA.RELAXATION_TIME: 
+      default:1
```
Comment (tuopuu) [inline simphony_metadata.yml]: add space

File: simphony_metadata.yml [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+      default: 1
+
+  CROSS_POWER_LAW_MODEL:
+    definition: Viscosity Cross power law model
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.INITIAL_VISCOSITY: 
+      default: 1e-3
+    CUBA.MAXIMUM_VISCOSITY: 
+      DEFAULT:1E-5   # nu_infiniy
```
Comment (tuopuu) [inline simphony_metadata.yml]: Also, lower case here: default: 1e-5

File: simphony_metadata.yml [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+      default: 0.5
+
+  BIRD_CARREAU_MODEL:
+    definition: The Bird-Carreau model
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.INITIAL_VISCOSITY: 
+      default: 1e-3
+    CUBA.MAXIMUM_VISCOSITY: 
+      DEFAULT:1E-5
```
Comment (tuopuu) [inline simphony_metadata.yml]: Lower-case.

**Comment by tuopuu** [comment]
@ahashibon, some minor typos and formatting errors, but apart from that, looks good to me.

File: simphony_metadata.yml [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+# If omitted, no default is assumed, and the user has to supply a
+# value.
+
+# scope: CUBA.USER or CUBA.SYSTEM.  if omitted, the CUBA.USER is
+# assumed by default
+
+# shape: as [i,j,] specifying that it is a sequence of entities of its
+# own type.  [i, j] means an array of i*j types.  [i, ] means the
+# sequence or array can have minimum of i elements. [, i] the array can
+# have any, maximum j elements. If omitted, a shape of [1] is assumed.
```
Comment (kitchoi) [inline simphony_metadata.yml]: > [i, ] means the sequence or array can have minimum of i elements, [, i] the array can have any, maximum j elements.

First of all, typo for the second part of the sentence (i and j).  Second of all, having this minimum and maximum flexibility is very strange.  Where is it needed?  And there is `shape: [1:]` further down this file.  What does that mean?

File: simphony_metadata.yml [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+  PHYSICS_EQUATION:
+   definition: SimPhoNy Physics Equation
+   parent: CUBA.MODEL_EQUATION
+
+  MATERIAL_RELATION_EQUATION:
+    definition: SimPhoNy Material Relation
+    parent: CUBA.MODEL_EQUATION
+    CUBA.MATERIAL:
+      scope: CUBA.USER
+      shape: [1,]
```
Comment (kitchoi) [inline simphony_metadata.yml]: Why is it necessary to have `shape: [1,]` here while other places have simply `[]`?

File: simphony_metadata.yml [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+      shape: [2]
+
+  DIRICHLET_BOUNDARY_CONDITION:
+    definition: Dirichlet boundary condition
+    parent: CUBA.CONDITION    
+    models: [CUBA.CONTINUUM]
+    CUBA.PHYSICS_VARIABLES: 
+      shape: [1:]
+    CUBA.MATERIAL: 
+      shape: [1:]
```
Comment (kitchoi) [inline simphony_metadata.yml]: What does this mean?

**Comment by kitchoi** [comment]
Except for a few questions I have above, the rest looks good to me.

File: simphony_metadata.yml (hunk 12 above)
Comment (ahashibon) [inline simphony_metadata.yml]: Thanks for noting the typo. The notation for the list may not be the best; this should be seen as an intermediate solution. There is a need to define values for some entities that are lists/sequence/series of a CUBA.  I first used a notation with a comma, but I found it later to be a bit inconsistent with standard notation, so I changed it to use colons and mixed up things a bit apparently.  Here is how the comment should look like (will be pushed shortly: 

`# shape: [i_min:i_max] specifies that the value is a sequence of entities of its own type.`
`# In each dimension, the first index is the minimum while the second maximum number of items allowed.`
`# similarly:`
`# shape: [i_min:i_max,j_min:j_max] specifies that the value is a multidimensional`
`# sequence of entities of its own type.  In each dimension, the first index`
`# is the minimum while the second maximum number of items allowed.`

Basically, we need a way to convey the information that the value is a list/array/sequence. This may still not be the best solution. Any ideas to enhance this are welcome.

File: simphony_metadata.yml (hunk 13 above)
Comment (ahashibon) [inline simphony_metadata.yml]: This is meant to convey that a material relation needs to have at least one material. Note that is should be `[1:]`.

File: simphony_metadata.yml (hunk 14 above)
Comment (ahashibon) [inline simphony_metadata.yml]: Please see previous comment.

**Comment by kitchoi** [comment]
The new push looks good to me!  Thanks @ahashibon!

**Comment by tuopuu** [comment]
Ok to merge?

**Comment by ahashibon** [comment]
will first create the initial Readme file with all the comments, and then it will be ok to merge.

**Comment by tuopuu** [comment]
@ahashibon, great! If you'll do a new push (for the ReadMe file) I can give it a quick review.

**Comment by ahashibon** [comment]
@tuopuu  please check it! Thanks!

File: simphony_metadata.yml [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+      default: 1
+      
+  HERSCHEL_BULKLEY_MODEL:
+    definition: Herschel-Bulkley model combines the effects of Bingham plastic and power-law behavior in a fluid
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.INITIAL_VISCOSITY: 
+      default: 1e-3
+    CUBA.RELAXATION_TIME: 
+      default: 1
```
Comment (kitchoi) [inline simphony_metadata.yml]: Where is the unit of relaxation time defined?

Comment (kitchoi) [inline simphony_metadata.yml]: I am mainly interested in knowing whether it must be an integer.

Comment (ahashibon) [inline simphony_metadata.yml]: It is not an integer ! 

SimPhoNy uses, currently, by default the SI system (MKS...). 
units should be introduced later. 
many new CUBA will have to be defined now in cuba.yml...

Comment (kitchoi) [inline simphony_metadata.yml]: Thanks.  Then it would be better to use `1.` instead of `1`, especially important for systems still using Python2.x

File: simphony_metadata.yml [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+    models: [CUBA.CONTINUUM]
+  
+  POWER_LAW_VISCOSITY_MODEL:
+    definition: Power Law model for a variable viscosity function that is limited by minimum and maximum values 
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.MAXIMUM_VISCOSITY: 
+      default: 1e-3
+    CUBA.MINIMUM_VISCOSITY:
+      default: 1E-5
```
Comment (tuopuu) [inline simphony_metadata.yml]: use lower-case format of exponential numbers: 1e-5

File: simphony_metadata.yml [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+      default: 1
+
+  CROSS_POWER_LAW_MODEL:
+    definition: Viscosity Cross power law model
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.INITIAL_VISCOSITY: 
+      default: 1e-3
+    CUBA.MAXIMUM_VISCOSITY: 
+      default: 1E-5   # nu_infiniy
```
Comment (tuopuu) [inline simphony_metadata.yml]: lower-case format also here: 1e-5, and fix the comment line typo.

File: simphony_metadata.yml [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+      default: 0.5
+
+  BIRD_CARREAU_MODEL:
+    definition: The Bird-Carreau model
+    parent: CUBA.RHEOLOGY_MODEL
+    models: [CUBA.CONTINUUM]
+    CUBA.INITIAL_VISCOSITY: 
+      default: 1e-3
+    CUBA.MAXIMUM_VISCOSITY: 
+      default: 1E-5
```
Comment (tuopuu) [inline simphony_metadata.yml]: 1e-5

**Comment by tuopuu** [comment]
@ahashibon, I saw just a few small typos, please see the line comments above. Otherwise it's good.

**Comment by ahashibon** [comment]
@tuopuu @kitchoi you guys are doing a great job! thanks. is ok to merge now?

**Comment by kitchoi** [comment]
Looks good to me! :+1:

**Comment by tuopuu** [comment]
:+1:
